File: sach/de3350.py
//DE3350.PY - Tran Huu Nam - hsgtin.vn - 2024
import math
def main():
    n, a, b, c, d, x, y, z, t, p, s = int(input()), 0, 0, 0, 0, 0, 0, 0, 0, 0, -1
    for d in range(1, int(math.sqrt(n)) + 1):
        for c in range(1, d + 1):
            for b in range(1, c + 1):
                p = n - d * d - c * c - b * b
                if p <= 0 or p > b * b:
                    continue
                a = int(math.sqrt(p))
                if a * a == p and s < a + b + c + d:
                    x, y, z, t = a, b, c, d
                    s = x + y + z + t
    if s < 0:
        print(-1)
    else:
        print(x, y, z, t)
if __name__ == "__main__":
    main()

File: sach/de3350.cpp
//DE3350.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
using namespace std;
int main()
{
    int n,a,b,c,d,x,y,z,t,p,s=-1;
    cin>>n;
    for (d=1; d*d < n; d++)
        for (c=1; c<=d; c++)
            for (b=1; b<=c; b++) {
                p=n-d*d-c*c-b*b;
                if (p<=0 || p>b*b) continue;
                a=sqrt(p);
                if (a*a==p && s<a+b+c+d) {
                    x=a;y=b;z=c;t=d;
                    s=x+y+z+t;
                }
            }
    if (s<0) cout<<-1;
    else cout<<x<<" "<<y<<" "<<z<<" "<<t;
    return 0;
}

